### PR TITLE
604 add doc upload page

### PR DESCRIPTION
**Pull request by marcellmueller**
Implements #604 

Also fixes missing form headers on both Operations and User Operator forms
Screenshot 2024-01-19 at 11 33 39 AM

Took me awhile to think of a good way to have different titles in the `MultiStepHeader` than what is in the form header (Statutory Declaration instead of the full `Statutory Declaration and Disclaimer`. I have added a `customHeaderTitles` option to the `MultiStepFormBase` that takes an array of strings and uses that for the titles instead of pulling from the form schema. 

I added a custom `jsxTitle` for the upload field itself since the design has bold text over a link to a download for the template. It's mostly commented out until we are given a link to the template download for the `statutory_declaration` field though it's styled and should be ready to go. I've created a follow up ticket as a reminder #645 

Screenshot 2024-01-19 at 10 24 17 AM

### CODE REVIEW

File: client/app/utils/jsonSchema/operations.ts [hunk 1]
```diff
@@ -415,12 +416,27 @@ const operationPage2: RJSFSchema = {
   ],
 };
 
+const operationPage3: RJSFSchema = {
+  type: "object",
+  title: "Statutory Declaration and Disclaimer",
+  // Uncomment this once documents are implemented
+  /*   required: ["statutory_declaration"], */
```
Comment (marcellmueller) [inline client/app/utils/jsonSchema/operations.ts:423]: Wasn't sure if I should leave this in for development purposes. Developers would still be able to get around this if we left it required by added a file even if nothing would happen on the backend.

File: client/app/components/form/titles/operationsTitles.tsx [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+  // Uncomment and replace the URL with the correct one when the template is ready in #645
+  // <>
+  //   Statutory Declaration
+  //   <br />
+  //   <span className="font-normal">
+  //     (Download <a href="TEMPLATE DOWNLOAD URL HERE">template</a>)
+  //   </span>
+  // </>
+  // Delete this line when the template is ready
+  <>Statutory Declaration</>
```
Comment (BCerki) [inline client/app/components/form/titles/operationsTitles.tsx:28]: Can we comment this back in and just have the link lead nowhere for now? It would be good for people to able to see this is done

**Review by BCerki — COMMENTED** [review]
Forms are looking pretty! 

I don't think this bug is related to this PR, but just in case it was a weird rebase thing, I'll mention here before making a ticket: https://www.loom.com/share/4192cf88269341118ffe9854aff419c9

File: client/app/components/form/MultiStepFormBase.tsx [hunk 3]
```diff
@@ -11,6 +11,8 @@ import MultiStepButtons from "./MultiStepButtons";
 interface MultiStepFormProps {
   baseUrl: string;
   cancelUrl: string;
+  // Optional array to override the default header titles
```
[the remark this answers is not in the document]
Comment (marcellmueller) [reply client/app/components/form/MultiStepFormBase.tsx:14]: Well it's in the AC and the design. I've had to push back against some major changes that were being asked for MVP and think I will keep this in. Also I think it's a good addition to the `MultiStepFormBase` and might make someones job easier in the future.

Of course if you ask Nan I will be happy to remove this though.

File: client/app/styles/rjsf/InlineFieldTemplate.tsx [hunk 4]
```diff
@@ -37,7 +37,10 @@ function InlineFieldTemplate({
   const error = rawErrors && rawErrors[0];
 
   // UI Schema options
-  const isLabel = uiSchema?.["ui:options"]?.label !== false;
+  const options = uiSchema?.["ui:options"] || {};
+  const jsxTitle = options?.jsxTitle as any;
+  const fieldLabel = jsxTitle ?? label;
+  const isLabel = options?.label !== false && fieldLabel;
```
[the remark this answers is not in the document]
Comment (marcellmueller) [reply client/app/styles/rjsf/InlineFieldTemplate.tsx:43]: Well the `FileWidget` here is wrapped in the `InlineFieldTemplate` so I think there is a misunderstanding and it seems odd to hardcode this into the `FileWidget`.

This is definitely a nice feature as it allows us to add bespoke titles to our fields if the designs require them like in this case. We are using it already in the `TitleOnlyFieldTemplate` and realistically it would be good to add in all field templates as required.

File: client/app/components/form/MultiStepFormBase.tsx [hunk 5]
```diff
@@ -61,10 +64,23 @@ const MultiStepFormBase = ({
...
+    );
+  }
 
   return (
     <>
-      <MultiStepHeader step={formSection} steps={formSectionTitles} />
+      <MultiStepHeader
+        step={formSection}
+        steps={isCustomTitles ? customHeaderTitles : formSectionTitles}
+      />
```
[the remark this answers is not in the document]
Comment (marcellmueller) [reply client/app/components/form/MultiStepFormBase.tsx]: Hmmm well I don't buy that it's somehow hard to follow but I don't feel strongly about it and have changed the prop to `customStepNames`.

[the remark this answers is not in the document]
Comment (BCerki) [reply client/app/components/form/MultiStepFormBase.tsx]: The prop name helps, thanks. In the future when responding to review comments, could you please tone down the skepticism? You've got a lot of experience with RJSF and you wrote this code, so it makes sense that it's going to be easier for you to follow than for someone else. I make hard-to-follow comments whenever it takes me a long time to review a PR relative to how big it is.

**Review by BCerki — APPROVED** [review]

File: client/app/components/form/MultiStepFormBase.tsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (marcellmueller) [reply client/app/components/form/MultiStepFormBase.tsx]: I'm so sorry, I am just not a huge fan of naming nits as what makes perfect sense to one person may not to another and in most libraries you need to double check what certain props actually do. Though I promise I will be better and more receptive in the future 🙇 

To expand on this a bit I think my main issue with this was the comment about not using `title` since RJSF uses it. Though to me using `custom` ... `title` in what is an RJSF component, replacing RJSF titles with custom ones is fine. Though I think `customStepNames` is nice and clear and I'm happy with the change.

File: client/app/styles/rjsf/TitleFieldTemplate.tsx [hunk 6]
```diff
@@ -0,0 +1,7 @@
+const TitleFieldTemplate = () => {
+  // RJSF isn't giving us access to 'ui:options' or 'ui:classNames' for customization
+  // so we're just going to return an empty fragment and handle the title in the field template
+  return <></>;
+};
+
+export default TitleFieldTemplate;
```
Comment (marcellmueller) [inline client/app/styles/rjsf/TitleFieldTemplate.tsx:7]: @BCerki what are your thoughts on this?

File: client/app/components/form/titles/operationsTitles.tsx [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+);
+
+export const StatutoryDeclarationUploadFieldTitle = (
+  <>
+    Statutory Declaration
+    <br />
+    <span className="font-normal">
+      (Download <a href="TEMPLATE DOWNLOAD URL HERE">template</a>)
+    </span>
+  </>
```
Comment (marcellmueller) [inline client/app/components/form/titles/operationsTitles.tsx:25]: Uncommented this thanks

**Review by BCerki — APPROVED** [review]

File: client/app/styles/rjsf/TitleFieldTemplate.tsx (hunk 6 above)
Comment (marcellmueller) [reply client/app/styles/rjsf/TitleFieldTemplate.tsx:7]: Thanks for taking another look. That `ui:options` is specific to the `multiple_operator_array` field though. I am not sure why they are sanitizing the main options even though they are supported in other templates.

Sounds good though, I still think this fix is better than hiding it with CSS selectors so I will merge it in even though it's imperfect.
